sdk/src/main/scala/coop/rchain/sdk/dag/merging/DagMergingLogic.scala

```diff
@@ -0,0 +1,378 @@
+  def partitionScope[D: Ordering](views: Seq[Set[D]]): Seq[Set[D]] = {
+    val r = LazyList.unfold(views) {
+      case (head @ seen) +: tail => (head, tail.map(_ -- seen)).some
+      case Seq()                 => none[(Set[D], List[Set[D]])]
+    }
+    r.toList
+  }
+
+  def computeRelationMap[D](
+      targetSet: Set[D],
+      sourceSet: Set[D],
+      relation: (D, D) => Boolean,
+      directed: Boolean
+  ): Map[D, Set[D]] =
+    targetSet.iterator
+      .flatMap(t => sourceSet.map((t, _)))
+      .foldLeft(mutable.Map.empty[D, Set[D]]) {
+        case (acc, (target, source)) =>
+          if (relation(target, source) && target != source) {
+            acc.update(source, acc.get(source).map(_ + target).getOrElse(Set(target)))
+            if (!directed)
+              acc.update(target, acc.get(target).map(_ + source).getOrElse(Set(source)))
+          }
+          acc
+      }
+      .toMap
+
+  /** Build conflicts map. */
+  def computeConflictsMap[D](
+      setA: Set[D],
+      setB: Set[D],
+      conflicts: (D, D) => Boolean
+  ): Map[D, Set[D]] =
+    computeRelationMap(setA, setB, conflicts, false)
+
+  /**
+    * Build dependency map.
+    * @param targetSet items that might depend on source set
+    * @param sourceSet items that target set might depend on
+    * @param depends dependency predicate
+    */
+  def computeDependencyMap[D](
+      targetSet: Set[D],
+      sourceSet: Set[D],
+      depends: (D, D) => Boolean
+  ): Map[D, Set[D]] =
+    computeRelationMap(targetSet, sourceSet, depends, true)
+
+  /** Compute branches of depending items. */
+  def computeBranches[D: Ordering](target: Set[D], depends: (D, D) => Boolean): Map[D, Set[D]] = {
+    val dependencyMap = computeDependencyMap(target, target, depends)
+    dependencyMap.foldLeft(dependencyMap) {
+      case (acc, (root, depending)) =>
+        val rootDependencies = acc.collect { case (k, v) if v.contains(root) => k }
+        if (rootDependencies.nonEmpty)
+          acc - root |+| rootDependencies.map(_ -> (depending + root)).toMap
+        else acc
+    } ++ (target -- dependencyMap.flatMap { case (k, v) => v + k }).map(_ -> Set.empty[D])
+  }
+
+  /**
+    * Compute branches of depending items that do not intersect.
+    * Partitioning done via [[partitionScopeBiggestFirst]].
+    */
+  def computeGreedyNonIntersectingBranches[D: Ordering](
+      target: Set[D],
+      depends: (D, D) => Boolean
+  ): Seq[Set[D]] = {
+    val concurrentRoots = computeBranches(target, depends)
+    val sorted =
+      concurrentRoots.toList.sortBy { case (k, v) => (-v.size, k) }.map { case (k, v) => v + k }
+    partitionScope(sorted).toList
+  }
+
+  /** Lowest fringe across number of fringes. */
+  def lowestFringe[B: Ordering](fringes: Set[Set[B]], height: B => Long): Set[B] = {
+    require(fringes.forall(_.nonEmpty), "Fringe has to have at least 1 element.")
+    require(fringes.nonEmpty, "Cannot compute lowest fringe on empty set.")
```

Comment:
Where is the resolution for the situation where fringe is empty like genesis block and blocks util first fringe?

Why this code cannot handle empty fringe and return empty fringe?

Reply:
I see, I'll allow fringe to be empty